diff --git a/kernel/init.c b/kernel/init.c
index a758c16ee0e..9fe4cf5c791 100644
--- a/kernel/init.c
+++ b/kernel/init.c
@@ -206,6 +206,8 @@ static void z_device_state_init(void)
 	}
 }
 
+__weak int tt_init_status;
+
 /**
  * @brief Execute all the init entry initialization functions at a given level
  *
@@ -245,6 +247,9 @@ static void z_sys_init_run_level(enum init_level level)
 		} else {
 			result = entry->init_fn();
 		}
+		if (result != 0 && tt_init_status == 0) {
+			tt_init_status = result;
+		}
 		sys_trace_sys_init_exit(entry, level, result);
 	}
 }
