diff --git a/drivers/i2c/i2c_ll_stm32.c b/drivers/i2c/i2c_ll_stm32.c
index 8601f0c03ba..5470bf81034 100644
--- a/drivers/i2c/i2c_ll_stm32.c
+++ b/drivers/i2c/i2c_ll_stm32.c
@@ -567,6 +567,8 @@ static const struct i2c_stm32_config i2c_stm32_cfg_##index = {					\
 			sizeof(i2c_timings_##index) / (sizeof(struct i2c_config_timing)),))	\
 	I2C_DMA_INIT(index, tx)									\
 	I2C_DMA_INIT(index, rx)									\
+	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32_i2c_v2),					\
+		(.transfer_timeout = K_USEC(DT_INST_PROP(index, transfer_timeout)),))		\
 };												\
 												\
 static struct i2c_stm32_data i2c_stm32_dev_data_##index = {					\
diff --git a/drivers/i2c/i2c_ll_stm32.h b/drivers/i2c/i2c_ll_stm32.h
index 5297bdc3a3f..0bc1a42cfd8 100644
--- a/drivers/i2c/i2c_ll_stm32.h
+++ b/drivers/i2c/i2c_ll_stm32.h
@@ -71,6 +71,7 @@ struct i2c_stm32_config {
 	struct stream tx_dma;
 	struct stream rx_dma;
 #endif /* CONFIG_I2C_STM32_V2_DMA */
+	k_timeout_t transfer_timeout;
 };
 
 struct i2c_stm32_data {
diff --git a/drivers/i2c/i2c_ll_stm32_v2.c b/drivers/i2c/i2c_ll_stm32_v2.c
index 5a1a4186e14..35c3250382a 100644
--- a/drivers/i2c/i2c_ll_stm32_v2.c
+++ b/drivers/i2c/i2c_ll_stm32_v2.c
@@ -32,8 +32,6 @@ LOG_MODULE_REGISTER(i2c_ll_stm32_v2);
 #include "i2c_ll_stm32.h"
 #include "i2c-priv.h"
 
-#define I2C_STM32_TRANSFER_TIMEOUT_MSEC  500
-
 #ifdef CONFIG_I2C_STM32_V2_TIMING
 /* Use the algorithm to calcuate the I2C timing */
 #ifndef I2C_STM32_VALID_TIMING_NBR
@@ -687,8 +685,7 @@ static int i2c_stm32_msg_write(const struct device *dev, struct i2c_msg *msg,
 	i2c_stm32_enable_transfer_interrupts(dev);
 	LL_I2C_EnableIT_TX(i2c);
 
-	bool is_timeout = (k_sem_take(&data->device_sync_sem,
-				      K_MSEC(I2C_STM32_TRANSFER_TIMEOUT_MSEC)) != 0);
+	bool is_timeout = (k_sem_take(&data->device_sync_sem, cfg->transfer_timeout) != 0);
 
 	if (is_timeout || data->cancelled) {
 		i2c_stm32_master_mode_end(dev);
@@ -750,8 +747,7 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
 	i2c_stm32_enable_transfer_interrupts(dev);
 	LL_I2C_EnableIT_RX(i2c);
 
-	bool is_timeout = (k_sem_take(&data->device_sync_sem,
-				      K_MSEC(I2C_STM32_TRANSFER_TIMEOUT_MSEC)) != 0);
+	bool is_timeout = (k_sem_take(&data->device_sync_sem, cfg->transfer_timeout) != 0);
 
 	if (is_timeout || data->cancelled) {
 		i2c_stm32_master_mode_end(dev);
@@ -845,7 +841,7 @@ static inline int msg_done(const struct device *dev,
 	const struct i2c_stm32_config *cfg = dev->config;
 	struct i2c_stm32_data *data = dev->data;
 	I2C_TypeDef *i2c = cfg->i2c;
-	int64_t start_time = k_uptime_get();
+	k_timepoint_t end_time = sys_timepoint_calc(cfg->transfer_timeout);
 
 	/* Wait for transfer to complete */
 	while (!LL_I2C_IsActiveFlag_TC(i2c) && !LL_I2C_IsActiveFlag_TCR(i2c)) {
@@ -856,8 +852,7 @@ static inline int msg_done(const struct device *dev,
 			/* proceed to issue stop */
 			break;
 		}
-		if ((k_uptime_get() - start_time) >
-		    I2C_STM32_TRANSFER_TIMEOUT_MSEC) {
+		if (sys_timepoint_expired(end_time)) {
 			return -ETIMEDOUT;
 		}
 	}
@@ -866,8 +861,7 @@ static inline int msg_done(const struct device *dev,
 		LL_I2C_GenerateStopCondition(i2c);
 		while (!LL_I2C_IsActiveFlag_STOP(i2c)
 		       && !data->cancelled) {
-			if ((k_uptime_get() - start_time) >
-			    I2C_STM32_TRANSFER_TIMEOUT_MSEC) {
+			if (sys_timepoint_expired(end_time)) {
 				return -ETIMEDOUT;
 			}
 		}
@@ -887,7 +881,7 @@ static int i2c_stm32_msg_write(const struct device *dev, struct i2c_msg *msg,
 	I2C_TypeDef *i2c = cfg->i2c;
 	unsigned int len = 0U;
 	uint8_t *buf = msg->buf;
-	int64_t start_time = k_uptime_get();
+	k_timepoint_t end_time = sys_timepoint_calc(cfg->transfer_timeout);
 
 	msg_init(dev, msg, next_msg_flags, slave, LL_I2C_REQUEST_WRITE);
 
@@ -902,8 +896,7 @@ static int i2c_stm32_msg_write(const struct device *dev, struct i2c_msg *msg,
 				return -EIO;
 			}
 
-			if ((k_uptime_get() - start_time) >
-			    I2C_STM32_TRANSFER_TIMEOUT_MSEC) {
+			if (sys_timepoint_expired(end_time)) {
 				return -ETIMEDOUT;
 			}
 		}
@@ -924,7 +917,7 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
 	I2C_TypeDef *i2c = cfg->i2c;
 	unsigned int len = 0U;
 	uint8_t *buf = msg->buf;
-	int64_t start_time = k_uptime_get();
+	k_timepoint_t end_time = sys_timepoint_calc(cfg->transfer_timeout);
 
 	msg_init(dev, msg, next_msg_flags, slave, LL_I2C_REQUEST_READ);
 
@@ -934,8 +927,7 @@ static int i2c_stm32_msg_read(const struct device *dev, struct i2c_msg *msg,
 			if (check_errors(dev, __func__)) {
 				return -EIO;
 			}
-			if ((k_uptime_get() - start_time) >
-			    I2C_STM32_TRANSFER_TIMEOUT_MSEC) {
+			if (sys_timepoint_expired(end_time)) {
 				return -ETIMEDOUT;
 			}
 		}
diff --git a/dts/bindings/i2c/st,stm32-i2c-v2.yaml b/dts/bindings/i2c/st,stm32-i2c-v2.yaml
index 80283d518b1..f861e8370ee 100644
--- a/dts/bindings/i2c/st,stm32-i2c-v2.yaml
+++ b/dts/bindings/i2c/st,stm32-i2c-v2.yaml
@@ -67,3 +67,13 @@ properties:
     description: |
       Names of the optional DMA channels. Expected values are "tx" for
       the TX channel and "rx" for the RX channel.
+
+  transfer-timeout:
+    type: int
+    default: 500000
+    description: |
+        I2C transfer timeout in microseconds.
+
+        This timeout is applied to message read and message write.
+        In non-interrupt mode the timeout may be applied again when
+        stop is transmitted.
