diff --git a/tests/drivers/dma/chan_link_transfer/src/test_dma.c b/tests/drivers/dma/chan_link_transfer/src/test_dma.c
index 78b52b77857..c83770587c5 100644
--- a/tests/drivers/dma/chan_link_transfer/src/test_dma.c
+++ b/tests/drivers/dma/chan_link_transfer/src/test_dma.c
@@ -24,19 +24,21 @@
 #define TEST_DMA_CHANNEL_0 (0)
 #define TEST_DMA_CHANNEL_1 (1)
 #define RX_BUFF_SIZE (48)
+#define DATA_ALIGNMENT DT_PROP(DT_NODELABEL(test_dma0), dma_buf_addr_alignment)
 
 #ifdef CONFIG_NOCACHE_MEMORY
-static __aligned(32) char tx_data[RX_BUFF_SIZE] __used
+static __aligned(DATA_ALIGNMENT) char tx_data[RX_BUFF_SIZE] __used
 	__attribute__((__section__(".nocache")));
 static const char TX_DATA[] = "It is harder to be kind than to be wise........";
-static __aligned(32) char rx_data[RX_BUFF_SIZE] __used
+static __aligned(DATA_ALIGNMENT) char rx_data[RX_BUFF_SIZE] __used
 	__attribute__((__section__(".nocache.dma")));
-static __aligned(32) char rx_data2[RX_BUFF_SIZE] __used
+static __aligned(DATA_ALIGNMENT) char rx_data2[RX_BUFF_SIZE] __used
 	__attribute__((__section__(".nocache.dma")));
 #else
-static const char tx_data[] = "It is harder to be kind than to be wise........";
-static char rx_data[RX_BUFF_SIZE] = { 0 };
-static char rx_data2[RX_BUFF_SIZE] = { 0 };
+static __aligned(DATA_ALIGNMENT) const char
+	tx_data[] = "It is harder to be kind than to be wise........";
+static __aligned(DATA_ALIGNMENT) char rx_data[RX_BUFF_SIZE] = { 0 };
+static __aligned(DATA_ALIGNMENT) char rx_data2[RX_BUFF_SIZE] = { 0 };
 #endif
 
 static void test_done(const struct device *dma_dev, void *arg, uint32_t id,
@@ -53,7 +55,7 @@ static int test_task(int minor, int major)
 {
 	struct dma_config dma_cfg = { 0 };
 	struct dma_block_config dma_block_cfg = { 0 };
-	const struct device *const dma = DEVICE_DT_GET(DT_NODELABEL(dma0));
+	const struct device *const dma = DEVICE_DT_GET(DT_NODELABEL(test_dma0));
 
 	if (!device_is_ready(dma)) {
 		TC_PRINT("dma controller device is not ready\n");
