diff --git a/dts/bindings/dma/dma-controller.yaml b/dts/bindings/dma/dma-controller.yaml
index 7df822886b1..111f93459aa 100644
--- a/dts/bindings/dma/dma-controller.yaml
+++ b/dts/bindings/dma/dma-controller.yaml
@@ -30,6 +30,7 @@ properties:
 
   dma-buf-addr-alignment:
     type: int
+    default: 32
     description: Memory address alignment requirement for DMA buffers used by the controller.
 
   dma-buf-size-alignment:
diff --git a/tests/drivers/dma/chan_blen_transfer/src/test_buffers.c b/tests/drivers/dma/chan_blen_transfer/src/test_buffers.c
index 3ec893f15c3..830190d6ce8 100644
--- a/tests/drivers/dma/chan_blen_transfer/src/test_buffers.c
+++ b/tests/drivers/dma/chan_blen_transfer/src/test_buffers.c
@@ -8,5 +8,11 @@
 
 #include "test_buffers.h"
 
-__aligned(32) char tx_data[TEST_BUF_SIZE] = "It is harder to be kind than to be wise........";
-__aligned(32) char rx_data[TEST_BUF_SIZE] = { 0 };
+#define DMA_ALIGN(tst_dma) DT_PROP(DT_NODELABEL(tst_dma), dma_buf_addr_alignment)
+
+#define DMA_INIT_DATA(tst_dma)                                                                     \
+	__aligned(DMA_ALIGN(tst_dma)) char tst_dma##_tx_data[TEST_BUF_SIZE] =                      \
+		"It is harder to be kind than to be wise........";                                 \
+	__aligned(DMA_ALIGN(tst_dma)) char tst_dma##_rx_data[TEST_BUF_SIZE] = {0};
+
+FOR_EACH(DMA_INIT_DATA, (), DMA_LIST);
diff --git a/tests/drivers/dma/chan_blen_transfer/src/test_buffers.h b/tests/drivers/dma/chan_blen_transfer/src/test_buffers.h
index d69bf914fdc..3c497dc581d 100644
--- a/tests/drivers/dma/chan_blen_transfer/src/test_buffers.h
+++ b/tests/drivers/dma/chan_blen_transfer/src/test_buffers.h
@@ -8,5 +8,11 @@
 
 #define TEST_BUF_SIZE (48)
 
-extern char tx_data[TEST_BUF_SIZE];
-extern char rx_data[TEST_BUF_SIZE];
+#define DMA_NAME(i, _) tst_dma##i
+#define DMA_LIST       LISTIFY(CONFIG_DMA_LOOP_TRANSFER_NUMBER_OF_DMAS, DMA_NAME, (,))
+
+#define DMA_DECLARE_DATA(tst_dma)                                                                  \
+	extern char tst_dma##_tx_data[TEST_BUF_SIZE];                                              \
+	extern char tst_dma##_rx_data[TEST_BUF_SIZE];
+
+FOR_EACH(DMA_DECLARE_DATA, (), DMA_LIST);
diff --git a/tests/drivers/dma/chan_blen_transfer/src/test_dma.c b/tests/drivers/dma/chan_blen_transfer/src/test_dma.c
index d5b2b7f7ad8..564e9d45de6 100644
--- a/tests/drivers/dma/chan_blen_transfer/src/test_dma.c
+++ b/tests/drivers/dma/chan_blen_transfer/src/test_dma.c
@@ -32,7 +32,8 @@ static void test_done(const struct device *dma_dev, void *arg,
 	}
 }
 
-static int test_task(const struct device *dma, uint32_t chan_id, uint32_t blen)
+static int test_task(const struct device *dma, uint32_t chan_id, uint32_t blen, char *tx_data,
+		     char *rx_data)
 {
 	struct dma_config dma_cfg = { 0 };
 	struct dma_block_config dma_block_cfg = { 0 };
@@ -60,8 +61,8 @@ static int test_task(const struct device *dma, uint32_t chan_id, uint32_t blen)
 		 dma->name, chan_id, blen >> 3);
 
 	TC_PRINT("Starting the transfer\n");
-	(void)memset(rx_data, 0, sizeof(rx_data));
-	dma_block_cfg.block_size = sizeof(tx_data);
+	(void)memset(rx_data, 0, TEST_BUF_SIZE);
+	dma_block_cfg.block_size = TEST_BUF_SIZE;
 #ifdef CONFIG_DMA_64BIT
 	dma_block_cfg.source_address = (uint64_t)tx_data;
 	dma_block_cfg.dest_address = (uint64_t)rx_data;
@@ -88,46 +89,49 @@ static int test_task(const struct device *dma, uint32_t chan_id, uint32_t blen)
 	return TC_PASS;
 }
 
-#define DMA_NAME(i, _) tst_dma##i
-#define DMA_LIST       LISTIFY(CONFIG_DMA_LOOP_TRANSFER_NUMBER_OF_DMAS, DMA_NAME, (,))
-
 #if CONFIG_DMA_TRANSFER_BURST16
 #define TEST_TASK(dma_name)                                                                        \
 	ZTEST(dma_m2m, test_##dma_name##_m2m_chan0_burst8)                                         \
 	{                                                                                          \
 		const struct device *dma = DEVICE_DT_GET(DT_NODELABEL(dma_name));                  \
-		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_0, 8) == TC_PASS));    \
+		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_0, 8,                  \
+					dma_name##_tx_data, dma_name##_rx_data) == TC_PASS));      \
 	}                                                                                          \
                                                                                                    \
 	ZTEST(dma_m2m, test_##dma_name##_m2m_chan1_burst8)                                         \
 	{                                                                                          \
 		const struct device *dma = DEVICE_DT_GET(DT_NODELABEL(dma_name));                  \
-		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_1, 8) == TC_PASS));    \
+		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_1, 8,                  \
+					dma_name##_tx_data, dma_name##_rx_data) == TC_PASS));      \
 	}                                                                                          \
                                                                                                    \
 	ZTEST(dma_m2m, test_##dma_name##_m2m_chan0_burst16)                                        \
 	{                                                                                          \
 		const struct device *dma = DEVICE_DT_GET(DT_NODELABEL(dma_name));                  \
-		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_0, 16) == TC_PASS));   \
+		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_0, 16,                 \
+					dma_name##_tx_data, dma_name##_rx_data) == TC_PASS));      \
 	}                                                                                          \
                                                                                                    \
 	ZTEST(dma_m2m, test_##dma_name##_m2m_chan1_burst16)                                        \
 	{                                                                                          \
 		const struct device *dma = DEVICE_DT_GET(DT_NODELABEL(dma_name));                  \
-		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_1, 16) == TC_PASS));   \
+		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_1, 16,                 \
+					dma_name##_tx_data, dma_name##_rx_data) == TC_PASS));      \
 	}
 #else
 #define TEST_TASK(dma_name)                                                                        \
 	ZTEST(dma_m2m, test_##dma_name##_m2m_chan0_burst8)                                         \
 	{                                                                                          \
 		const struct device *dma = DEVICE_DT_GET(DT_NODELABEL(dma_name));                  \
-		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_0, 8) == TC_PASS));    \
+		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_0, 8,                  \
+					dma_name##_tx_data, dma_name##_rx_data) == TC_PASS));      \
 	}                                                                                          \
                                                                                                    \
 	ZTEST(dma_m2m, test_##dma_name##_m2m_chan1_burst8)                                         \
 	{                                                                                          \
 		const struct device *dma = DEVICE_DT_GET(DT_NODELABEL(dma_name));                  \
-		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_1, 8) == TC_PASS));    \
+		zassert_true((test_task(dma, CONFIG_DMA_TRANSFER_CHANNEL_NR_1, 8,                  \
+					dma_name##_tx_data, dma_name##_rx_data) == TC_PASS));      \
 	}
 #endif
 
